### ISSUE TITLE
set_requires_grad(net_D, False) when updating the generator?

### ISSUE THREAD

**Issue by trougnouf** [junyanz/pytorch-CycleGAN-and-pix2pix#651]
I'm a bit confused about the use of set_requires_grad(net_D, False) when training the generator (ie pix2pix_model.py), wouldn't the generator need the discriminator's gradient to backpropagate and learn? Otherwise it seems the generator would only get the loss values.

**Comment by junyanz** [comment]
It's a small speedup trick. set_requires_grad=False will stop calculating gradients for the discriminator while we update the generator. It can save some time and memory. Generators will get the gradients.

**Comment by trougnouf** [comment]
Thank you!

**Comment by micklexqg** [comment]
@junyanz, but when i load the pretrained netD for resuming training,   error occurs:
loaded state dict contains a parameter group that doesn't match the size of optimizer's group.
when i do not set set_requires_grad, this error does not occur.

**Comment by junyanz** [comment]
Not sure if "set_requires_grad` will affect the parameter group.

**Comment by tanluDIMA** [comment]
> It's a small speedup trick. set_requires_grad=False will stop calculating gradients for the discriminator while we update the generator. It can save some time and memory. Generators will still get the gradients.

Sorry to come back to this but it's still not clear to me. Does this mean that the gradients of the discriminator weights (which are required to backpropagate to the generator) will still be calculated but not stored in the memory? 
A second question, why optimize the generator before the discriminator (which opposes the optimization flow of the original GAN algorithm ?)

**Comment by junyanz** [comment]
1. The gradients of the discriminator are not calculated. Please see this [doc](https://pytorch.org/docs/stable/notes/autograd.html) for more details. 
2. The order does not matter in practice.

**Comment by tanluDIMA** [comment]
> 1. The gradients of the discriminator are not calculated. Please see this [doc](https://pytorch.org/docs/stable/notes/autograd.html) for more details.
> 2. The order does not matter in practice.

Thank you very much for the prompt reply! I have read through the documentation which is helpful to clear my minds. All along my mistake (please advise if I still did not get this part right) was not realizing that it is the weights, instead of the gradients of the weights, of the discriminator that are required in order to backpropagate to the generator. By freezing the discriminator, anything that is dependent on the gradient of the weights of the discriminator will be cut off from the graph, this however does not obstruct the trace back to the weights of the generator.
